### PR TITLE
MooseDocs re-factored to be entirely markdown.Extension based

### PR DESCRIPTION
**Pull request by aeslaughter**
This is a bunch of code that I have been working with on and off over the past few months and is motivated by upcoming documentation needs for Mastodon and SQA requirements. 

The key features of these changes include:

1. python/MooseDocs becomes pylint clean
1. Adds large number of tests
1. MooseDocs is converted to my completely markdown.Extension based
1. LaTeX conversion re-factored to be more robust
1. Adds the ability to make arbitrary reference floats (e.g., figures, tables, listings, etc.)

(refs #8329)

### CODE REVIEW

**Comment by aeslaughter** [comment]
I think this is ready to go and just to give some context to this change set, the main driver was simply to make the MooseDocs code design more uniform, as mentioned above nearly everything is now down via markdown Extensions and the extensions where divided up into multiple extensions to make them more manageable. 

With this change I updated the MooseDocs documentation: http://mooseframework.org/docs/PRs/9050/site/utilities/moose_docs/moose_markdown/index.html

To give an idea of where all the additions come from here are some numbers, a majority of the added lines come from new tests.

docs/content: 102 files changed, 1187 insertions(+), 828 deletions(-)
python/MooseDocs: 238 files changed, 8342 insertions(+), 3371 deletions(-)
python/MooseDocs/tests:  168 files changed, 2920 insertions(+)

**Comment by aeslaughter** [comment]
Mastodon will fail, but I have a patch.

**Comment by aeslaughter** [comment]
@sapitts @dschwen You are probably want to spend time looking through the changes I made: http://mooseframework.org/docs/PRs/9050/site/utilities/moose_docs/index.html

**Comment by aeslaughter** [comment]
@sapitts FYI, I have a patch that will update BISON

**Comment by sapitts** [comment]
Thank you so very much for the Bison patch!

**Comment by dschwen** [comment]
Can I haz patch for Marmot?

File: python/MooseDocs/html2latex/moose_elements.py [hunk 1]
```diff
@@ -1,322 +1,285 @@
...
+#        self._path = None
+#        return False
+#
+#    def convert(self, tag, content):
+#        url = '{}/{}'.format(self._site, self._path)
+#        return '\\href{%s}{%s}' % (url, content)
+#
+#    def preamble(self):
+#        return ['\\usepackage{hyperref}']
+
```
Comment (brianmoose) [inline python/MooseDocs/html2latex/moose_elements.py]: Do you need all this commented out stuff?

File: python/MooseDocs/html2latex/moose_elements.py [hunk 2]
```diff
@@ -1,322 +1,285 @@
...
+#    """
+#    Produce and error that videos are not supported.
+#    """
+#    name = 'video'
+#    def convert(self, tag, content):
+#        LOG.warning('The !video syntax is not currently supported for latex/pdf output.')
+#        return '\\admonition[error-title][error]{ERROR: Un-supported Markdown!}{!video}'
+#    def preamble(self):
+#        return admonition_preamble()
+#
```
Comment (brianmoose) [inline python/MooseDocs/html2latex/moose_elements.py]: More commented out stuff...

**Review by brianmoose — DISMISSED** [review]
Not something this PR needs to address but the "Create an App" page probably needs to be updated to reflect the use of the new script rather than "Fork the Stork".

FAQ is empty...

**Comment by aeslaughter** [comment]
@dschwen I will give a patch to Marmot when I add the ability to include submodule documentation, which I am working on now.

**Comment by aeslaughter** [comment]
@brianmoose The comment out code is working-in-progress with latex conversion, I can remove it.

**Comment by permcody** [comment]
This says ready to merge but it also failed... Is that OK?

**Comment by aeslaughter** [comment]
@permcody  Mastodon will fail, but I have a patch.

**Comment by aeslaughter** [comment]
@brianmoose Can you enable pylint on the python/MooseDocs directory once this goes in?

**Comment by brianmoose** [comment]
Yeah, adding python/MooseDocs to pylint is ready to go in.

**Review by brianmoose — APPROVED** [review]

**Pull request merged**